src/features/weak-references.md

```diff
@@ -54,104 +54,133 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
 
 ```js
-const cache = new Map();
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    this.socket = socket;
+    this.listener = (ev) => { this.events.push(ev); };
+    socket.addEventListener('message', this.listener);
+  }
 
-function getImageCached(name) {
-  const ref = cache.get(name);
-  if (ref !== undefined) {
-    const deref = ref.deref();
-    if (deref !== undefined) return deref;
+  dispose() {
+    this.socket.removeEventListener('message', this.listener);
   }
-  const image = performExpensiveOperation(name);
-  const wr = new WeakRef(image);
-  cache.set(name, wr);
-  return image;
+
+  // ...
 }
 ```
 
-But there’s still a problem here: the `Map` still holds on to the `name` strings forever, because those are the keys in the cache. Ideally, those strings would be removed too. The `WeakRef` proposal has a solution for this as well! With the new `FinalizationRegistry` API, we can register a callback to run when the garbage collector zaps a registered object. Such callbacks are known as _finalizers_.
+The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. Because the application behavior doesn't depend on the event listener of this diagnostic class, and that the listener is expensive in terms of memory use but not in computation, we really want the listener's lifetime to be logically tied to the `MovingAvg` instance. If its lifetime were tied to the `MovingAvg` instance, then `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.
 
-:::note
-**Note:** The finalization callback does not run immediately after garbage-collecting the image object. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
-:::
-
-Here, we register a callback to remove keys from the cache when the image objects are garbage-collected:
+`WeakRef`s make it possible to solve the dilemma by creating a _weak reference_ to the actual event listener, and then wrapping that `WeakRef` in an outer event listener. This way, the garbage collector can clean up the actual event listener and the memory that it holds alive, like the `events` array.
 
 ```js
-const cache = new Map();
-
-const finalizationRegistry = new FinalizationRegistry((name) => {
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    const listener = (ev) => { this.events.push(ev); }
+    const weakRef = new WeakRef(listener);
+    const listenerWrapper = (ev) => {
+      const deref = weakRef.deref();
+      if (deref !== undefined) deref(ev);
+    });
+    socket.addEventListener("message", listenerWrapper);
```

Comment:
Doesn't this still leak the listenerWrapper? The socket doesn't know that the weakref died, and doesn't remove listenerWrapper.

In general, aren't the weak refs "off by one layer" here? The eventlistener list should be a list of weakrefs, so that we can clean up the eventlisteners which are no longer relevant.

Or did you mean the cleanup would be done by listenerwrapper when it gets an event and it notices its weakref has expeired?

Reply:
It does indeed leak the listenerWrapper, which is the point of the subsequent finalization registry note.

Reply:
Ah, right, I was confused by the "No more dispose!" below which I took to mean "dispose is not needed now, yay". Which is confusing, like, it's not needed in the first place if leaking is OK...

Reply:
(It might be less confusing to still have the dispose here because we still need to remove the event listener, like in the previous version. But I realize it is probably way too late to fix this, so pls feel free to ignore this comment.)

Reply:
Removed the comment that said "No more dispose!"